src/stim/io/measure_record_reader.cc

```diff
@@ -312,63 +341,71 @@ size_t MeasureRecordReaderFormatR8::read_bytes(PointerRange<uint8_t> data) {
 }
 
 bool MeasureRecordReaderFormatR8::read_bit() {
-    if (position >= bits_per_record) {
-        throw std::out_of_range("Attempt to read past end-of-record");
+    if (!buffered_0s && !buffered_1s) {
+        buffer_data();
     }
-    if (generated_1s < run_length_1s) {
-        ++generated_1s;
-        ++position;
-        return true;
-    }
-    if (generated_0s < run_length_0s) {
-        ++generated_0s;
-        ++position;
+    if (buffered_0s) {
+        buffered_0s--;
+        position++;
         return false;
-    }
-    if (!update_run_length()) {
-        throw std::out_of_range("Attempt to read past end-of-file");
-    } else {
-        ++generated_1s;
-        ++position;
+    } else if (buffered_1s) {
+        buffered_1s--;
+        position++;
         return true;
+    } else {
+        throw std::invalid_argument("Read past end-of-record.");
     }
 }
 
 bool MeasureRecordReaderFormatR8::next_record() {
+    while (!is_end_of_record()) {
+        read_bit();
+    }
     position = 0;
-    return false;
+    have_seen_terminal_1 = false;
+    int i = fgetc(in);
+    if (i == EOF) {
+        return false;
+    }
+    ungetc(i, in);
+    return true;
 }
 
 bool MeasureRecordReaderFormatR8::is_end_of_record() {
-    if (position >= bits_per_record) {
-        return true;
-    }
-    if (generated_0s < run_length_0s) {
-        return false;
-    }
-    if (generated_1s < run_length_1s) {
-        return false;
-    }
-    return !update_run_length();
+    return position == bits_per_record && have_seen_terminal_1;
 }
 
-bool MeasureRecordReaderFormatR8::update_run_length() {
-    int r = getc(in);
-    if (r == EOF) {
-        return false;
+void MeasureRecordReaderFormatR8::buffer_data() {
+    assert(buffered_0s == 0);
+    assert(buffered_1s == 0);
+    if (is_end_of_record()) {
+        throw std::invalid_argument("Attempted to read past end-of-record.");
     }
-    run_length_0s = 0;
-    while (r == 0xFF) {
-        run_length_0s += 0xFF;
+
+    // Count zeroes until a one is found.
+    int r;
+    do {
         r = getc(in);
+        if (r == EOF) {
+            throw std::invalid_argument("r8 data hit end-of-file without seeing a 1 encoded at end of data.");
+        }
+        buffered_0s += r;
+    } while (r == 0xFF);
+    buffered_1s = 1;
+
+    // Check if the 1 is the one at end of data.
+    size_t total_data = position + buffered_0s + buffered_1s;
+    if (total_data == bits_per_record) {
+        if (getc(in) != 0) {
+            throw std::invalid_argument("r8 data ending in an encoded 1 didn't include a 0x00 byte encoding the extra 1 at end of data.");
+        }
+        have_seen_terminal_1 = true;
+    } else if (total_data == bits_per_record + 1) {
+        have_seen_terminal_1 = true;
+        buffered_1s = 0;
+    } else if (total_data > bits_per_record + 1) {
+        throw std::invalid_argument("r8 data encoded a jump past the expected end of encoded data.");
```

Comment:
Out of curiosity: Was the change in logic here necessary to implement the missing support for multiple records in R8? I agree that the fact that `generated_1s` was an integer might seem a little odd (since the variable can only ever be =0 or =1, unlike `generated_0s`), but that choice did simplify the logic! The new logic is more complicated. In particular, it requires the `have_seen_terminal_1` switch which leads to more branches that could hurt performance and make it harder to have good test coverage.

Reply:
Oh, was the old logic accounting for the terminal 1? I wasn't able to see that it was doing that so I just ripped it out and did what you see here.

Reply:
Well, I might have had a bug, but I did recognize the need to do so and intended to account for it. There was some related subtleties around (re-)initializing run_length_1s related to that goal.

Anyway, I was just curious. I don't object to the new logic, especially since it handles multiple records (unlike the old logic which was written under the wrong assumption that R8 doesn't have multiple records).